Leverage twistlock api in checkov

Eliran-Turgeman opened the pull request:
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Changed checkov to leverage the twistlock wrapper API.

Package scanning won't require installing twistcli client and will store results in the cache.
Image scanning will still require to install the twistcli client but will utilize the cache as well.

itai1357 commented on checkov/common/bridgecrew/vulnerability_scanning/image_scanner.py:
```diff
@@ -109,5 +114,29 @@ def scan(self, docker_image_id: str, dockerfile_path: str, skip_extract_image_na
+                    response.json()["outputData"]
+                )
+            )
+        return {}
```
i suggest one of the next here:
1. writing a comment here for describing that output_type == "Empty" in case it reaches to line 139, 
2.  adding third 'elif' (output_type == "Empty") statement and another else
in the else segment, just raise an error, so we are monitored in case there is new output_type which are not known here. 
basically, it is up to you
@nimrodkor what I suggest as option 2 is good practice or too much according to your opinion

nimrodkor replied:
Would log an info log in this case, something like `Got an empty result for file {file_name}`.
1 - doesn't help the average user
2 - we can't monitor the cli runs on user computers... So we'll have to let them know, gently, so they can reach out. raising an error will crash the scan, and we don't want that if we did something wrong...

itai1357 commented on checkov/sca_package/scanner.py:
```diff
@@ -2,103 +2,110 @@
-        process = await asyncio.create_subprocess_shell(
-            command, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+
+        headers = merge_dicts(
```
maybe create a method "get_header" in this class. we have the same logic in line 22.
if it isn't to hard, the best is create some "global" function "get_header" (which can be part of 'bc_integration') and than we can use also it instead the header-creation in image_scanner.py:118 (and apparently in a lot of other places), but it may be more complicated, so let's start do it in this class

itai1357 commented on checkov/sca_package/scanner.py:
```diff
@@ -2,103 +2,110 @@
+        )
+        response = requests.Response()
+
+        while current_state != desired_state:
```
add some limitation. (e.g: after 2 minutes, abort and log it)

nimrodkor replied:
Agree with @itai1357 , Let's limit it to 60 seconds please

itai1357 replied:
@Eliran-Turgeman just making sure you fixed it

Eliran-Turgeman replied:
Yeah done, refer to lines 95 - 100

itai1357 commented on tests/sca_package/test_scanner.py:
```diff
@@ -1,137 +1,102 @@
-        "checkov.common.bridgecrew.vulnerability_scanning.integrations.package_scanning.package_scanning_integration.download_twistcli",
-        side_effect=integration_mock,
-    )
+def assert_equal_dicts(a, b):
```
if we use a={'a': 1}, b= {'a': 2} we got False-Positive
I guess that you have any assumption regarding the dicts 'a' and 'b'. please write them in the documentation (or as a comment) of the function

nimrodkor replied:
Need to fix this still @Eliran-Turgeman

itai1357 commented on tests/sca_package/test_scanner.py:
```diff
@@ -1,137 +1,102 @@
 
-
-def test_setup_twistcli_not_exists(mocker: MockerFixture, tmp_path: Path):
+@responses.activate
```
nice decorator! Big Like

itai1357 commented on tests/sca_package/test_scanner.py:
```diff
@@ -1,137 +1,102 @@
-@pytest.mark.asyncio
-async def test_run_scan(mocker: MockerFixture, tmp_path: Path, mock_bc_integration, scan_result):
+@responses.activate
+def test_run_scan_fail_on_scan(mock_bc_integration):
```
nice test!!

itai1357 reviewed (commented):
Great work and awesome tests! some small comments

nimrodkor commented on checkov/common/bridgecrew/platform_integration.py:
```diff
@@ -21,6 +21,7 @@
 from tqdm import trange
 from urllib3.exceptions import HTTPError, MaxRetryError
 
+from checkov.common.bridgecrew.bc_source import get_source_type
```
@Eliran-Turgeman necessary import?

nimrodkor commented on checkov/sca_package/scanner.py:
```diff
@@ -2,103 +2,110 @@
+        self.base_url = bc_integration.api_url
+        self.headers = merge_dicts(
+            get_default_get_headers(bc_integration.bc_source, bc_integration.bc_source_version),
+            {"Authorization": bc_integration.get_auth_token()},
```
@Eliran-Turgeman This might be a time-limited token, and so should be computed as close to the request as possible. It contains a cache by itself, so don't worry calling it too many times :)

nimrodkor reviewed (commented):
Very nice work @Eliran-Turgeman !
Need some cleaning up around the requests to reuse code as much as possible

Eliran-Turgeman commented:
Thanks for your review @itai1357 @nimrodkor!
I addressed all your comments, the PR is ready for a second review :)

There's a small addition i'll have to make regarding the input validation on docker image id (make it in a sha256 format) which I am testing at the moment

gruebel commented on checkov/common/bridgecrew/vulnerability_scanning/image_scanner.py:
```diff
@@ -109,5 +113,28 @@ def scan(self, docker_image_id: str, dockerfile_path: str, skip_extract_image_na
             return 1
 
+    def get_scan_results_from_cache(self, image_id: str) -> Dict[str, Any]:
+        image_id_encode = quote_plus(':'.join(['sha256', image_id]))
```
```suggestion
        image_id_encode = quote_plus(f"sha256:{image_id}")
```
better to use a f-string literal here, also looks nicer 😉

gruebel commented on checkov/common/bridgecrew/vulnerability_scanning/image_scanner.py:
```diff
@@ -109,5 +113,28 @@ def scan(self, docker_image_id: str, dockerfile_path: str, skip_extract_image_na
+        elif output_type == "Result":
+            return json.loads(
+                decompress_file_gzip_base64(
+                    response.json()["outputData"]
```
```suggestion
                    response_json["outputData"]
```
no need to do it again

gruebel commented on checkov/common/util/file_utils.py:
```diff
@@ -1,3 +1,36 @@
+        # 2. decode base64 data
+        # 3. wrap decoded binary data with BytesIO to enable reading
+        # 4. gunzip compressed data
+        decoded_utf8 = bytes(compressed_file_body, "utf-8")
+        decoded_base64 = base64.b64decode(decoded_utf8)
```
```suggestion
        base64_bytes = compressed_file_body.encode("utf-8")
        decoded_base64 = base64.b64decode(base64_bytes)
```
to keep similar to the compress function code

gruebel commented on checkov/sca_package/scanner.py:
```diff
@@ -5,116 +5,102 @@
     def __init__(self) -> None:
-        self.twistcli_path = Path(bridgecrew_dir) / TWISTCLI_FILE_NAME
-
-    def setup_twictcli(self) -> None:
-        try:
-            if self.should_download():
-                if not os.path.exists(bridgecrew_dir):
-                    os.makedirs(bridgecrew_dir)
-                self.cleanup_twictcli()
-                package_scanning_integration.download_twistcli(self.twistcli_path)
-        except Exception:
-            logging.error("Failed to setup twictcli for package scanning", exc_info=True)
-            raise
```
just to make sure, do we remove already what Noa implemented today?

nimrodkor replied:
We spoke, it needs to be implemented on the sca_image runner - but on the sca_package runner it will become obsolete soon

gruebel reviewed (commented):
nice work 💪

nimrodkor commented on checkov/sca_package/scanner.py:
```diff
@@ -5,116 +5,102 @@
-        # read and delete the report file
-        scan_result = json.loads(output_path.read_text())
-        output_path.unlink()
+            time.sleep(SLEEP_DURATION)
+            total_sleeping_time += SLEEP_DURATION
```
@Eliran-Turgeman let's put this line after the `if total_sleeping_time > MAX_SLEEP_DURATION:` to avoid waiting without making a call

itai1357 commented on checkov/common/bridgecrew/platform_integration.py:
```diff
@@ -641,5 +641,16 @@ def repo_matches(self, repo_name):
         # matches xyz_org/repo or org/repo (where xyz is the BC org name and the CLI repo prefix from the platform)
         return re.match(re.compile(f'^(\\w+_)?{self.repo_id}$'), repo_name) is not None
 
+    def get_default_headers(self, request_type: str) -> dict:
```
nice!

itai1357 commented on checkov/sca_image/runner.py:
```diff
@@ -52,8 +62,8 @@ def scan(
             raise
 
-    @staticmethod
     async def execute_scan(
```
why did you choose to put the method "execute_scan" in the runner rather than in "image_scanner.py"?

Eliran-Turgeman replied:
It was originally in the runner, I haven't moved it.
Should I move it to the image_scanner?

itai1357 commented on checkov/sca_image/runner.py:
```diff
@@ -73,8 +83,26 @@ async def execute_scan(
         scan_result: Dict[str, Any] = json.loads(output_path.read_text())
+
+        # upload results to cache
+        image_id_sha = ':'.join(['sha256', image_id])
```
I suggest to use here also 'quote_plus'
`image_id_sha = quote_plus(f"sha256:{image_id}")`
(as you did in the image_scanner.py)

Eliran-Turgeman replied:
Will modify to f-string, but I think the usage of `quote_plus` isn't needed and might lead to undesired behavior since the id is passed in the body and not the url itself so it won't be decoded

itai1357 commented on tests/sca_package/test_runner.py:
```diff
@@ -134,59 +134,39 @@ def test_prepare_and_scan(mocker: MockerFixture, scan_result):
     assert runner._code_repo_path == EXAMPLES_DIR
 
 
-def test_prepare_and_scan_sca_package_scan_disabled(mocker: MockerFixture, scan_result):
```
why did you delete it?

Eliran-Turgeman replied:
This test is verifying that the feature flag around sca package scanning is working and returns an empty result.
Since we wanted to delete the feature flag around this feature, I deleted the test, otherwise it would fail.

itai1357 reviewed (commented):
Awesome. some small questions. if the reasons are good, tell me and I will apptove

itai1357 reviewed (approved):
Great work!
